feat: add support for ExpressLRS optional arming method

mha1 opened the pull request:
This PR adds support for ExpressLRS/ExpressLRS#3008

Changes:
- adds UI to select Arming mode Switch and to set a switch (logical, physical or any other suitable source)
- sends requested arming state in extended CRSF message 0x16 in arming mode Switch

For more details please refer to ExpressLRS PR [#3008](https://github.com/ExpressLRS/ExpressLRS/pull/3008)

ulfhedlund commented:
Great new functionality!
SE translation: 
#define TR_SF_ARM "Aktivera"
#define TR_ARMING_MODE  "Aktivering"

ajjjjjjjj commented:
This seems to be designed as CRSF specific, but is only tied to ELRS. How about adding "ELRS" to all labels to improve UX?
Also this could be done in lua by using `getSwitch*()` and/or `getLogicalSwitchValue()` functions.

I'm not EdgeTX advocate nor involved as developer/UX designer but I don't like how ELRS configuration must be done in many places (lua, specific module configuration, specific special function) where all of this can be kept in lua and maintained by original team.

I know historicaly MPM is also integrated in EdgeTX but because there was no lua by the time and other constraints. Now, there is better way (LUA) so i don't think it is the way to go nor user friendly.

I'm not against your change, I just see other path that could be considered by project maintainers to make both projects better.

philmoz commented on radio/src/datastructs_private.h:
```diff
@@ -995,8 +996,6 @@ PACK(struct RadioData {
   uint8_t invertLCD:1;          // Invert B&W LCD display
   NOBACKUP(uint8_t spare:3 SKIP);
-#else
-  NOBACKUP(uint8_t spare:4 SKIP);
```
Why has this been deleted? Removing it breaks the alignment.

mha1 replied:
Thanks for pointing that out. Not intentional. My excuse: fat fingers.

I'll fix that with together with the missing translation

philmoz commented on radio/src/datastructs_private.h:
```diff
@@ -693,11 +693,12 @@ PACK(struct ModelData {
   uint8_t   disableTelemetryWarning:1;
   uint8_t   showInstanceIds:1;
   uint8_t   checklistInteractive:1;
+  uint8_t   crsfArmingMode:1;
```
Why is this in the Model data and not the Module data?

mha1 replied:
Consider the rather rare corner case where an internal and external module are active at the same time e.g. to build a 900Mhz/2.4GHz redundancy. Kinda nonsense with GemX being around but we currently don't reject a dual module setup even if telemetry is a problem.

What we do currently for this corner case wrt arming is to have both modules listening to CH5. The same CH5 data is sent to the internal and external module. Both modules will arm/disarm synchronously. So arming is already effectively a model property rather than a module property.

We could make it module specific or which I'd prefer (gut feeling) leave it model specific meaning in the case of two active modules both would be armed/disarmed synchronously in both arming modes

philmoz commented on radio/src/translations/en.h:
```diff
@@ -65,6 +65,8 @@
 #define TR_VVARIOCENTER                "Tone","Silent"
 #define TR_CURVE_TYPES                 "Standard","Custom"
 
+#define TR_CRSF_ARMING_MODES           "Channel","Function"
```
Should 'Channel' option be 'Channel5' to make the intent clearer?
Also this is missing in other language files.

philmoz commented:
Why does this need a special function? This seems overly complicated to me.

Wouldn't it be simpler and easier for users to just use a switch that the user can select.
Then the options would be 'Channel5' and 'Switch'.
If 'Switch' is selected then the user can use the existing switch selection UI to pick which switch to use for arming.
There looks to be plenty of room in the ModuleData structure to put both the arming mode flag and the selected arming switch in the 'crsf' section.

pfeerick commented:
> This seems to be designed as CRSF specific, but is only tied to ELRS. How about adding "ELRS" to all labels to improve UX?
> Also this could be done in lua by using `getSwitch*()` and/or `getLogicalSwitchValue()` functions.


I guess if the same mechanism as is used to display the bind option only for ELRS that could be done. Especially since this is a ELRS only option. For configuration, sure Lua is fine, but I don't like it for critical mixer tasks - i.e. I would want to be certain this wouldn't cause the link to be dropped if Lua is terminated. 


> Why does this need a special function? This seems overly complicated to me.
> 
> Then the options would be 'Channel5' and 'Switch'.
> If 'Switch' is selected then the user can use the existing switch selection UI to pick which switch to use for arming.

As long as you keep in mind it may not even need to be a physical switch but could be a LS (or even constant 😀), that would be a much better UX, as well as it bit more self-documenting. .

mha1 commented:
> Why does this need a special function? This seems overly complicated to me.
> 
> Wouldn't it be simpler and easier for users to just use a switch that the user can select. Then the options would be 'Channel5' and 'Switch'. If 'Switch' is selected then the user can use the existing switch selection UI to pick which switch to use for arming. There looks to be plenty of room in the ModuleData structure to put both the arming mode flag and the selected arming switch in the 'crsf' section.

Actually that's a good idea idea if "Switch" can be a logical switch too which I think it does, doesn't it?

I agree on "Channel" vs "Channel5". maybe even better "AUX1(CH5)" because ExpressLRS/BF mostly use the term AUX1 if that fits the small b&w screens too.

pfeerick commented:
AUX1 is inav/bf notation though? What about ardupilot, sbus etc? May be better to not have non-ELRS specific verbiage. After all, we don't use any other BF/INAV specific.

mha1 commented:
> AUX1 is inav/bf notation though? What about ardupilot, sbus etc? May be better to not have non-ELRS specific verbiage. After all, we don't use any other BF/INAV specific.

Hah, same thought about "Switch". And you're right about ON/OFF too. Maybe I'm wrong but the reason why I chose SF Arm is the fact all the logic for activation/deactivation is already there and can be used. Having a "Switch" selection in the CRSF UI just defines the switch. If "Switch" is more than physical switch, ie more like "Trigger" the logic to determine the status needs to be done. It'd be like a hidden internal SF.

AUX1 is also ExpressLRS lingo. Open the online documentation at expresslrs.org and search for AUX1. But really, I'm good with Channel5 or CH5 (if Channel5 doesn't fit the b&w screens).

Sorry about the missing translation. I missed that one. Will update everything once we agree on the UI. I like the idea to keep it all to the CRSF UI.

pfeerick commented:
>  activation/deactivation is already there and can be used. Having a "Switch" selection in the CRSF UI just defines the switch. If "Switch" is more than physical switch, ie more like "Trigger" the logic to determine the status needs to be done. It'd be like a hidden internal SF.

True... but all you really need access to is the "trigger" part... don't need to define which SF since the context defines that, and don't need to be able to disable/enable it... so would be better to integrate it in the module setup entirely (or forget the whole thing, and do it via Lua... yeah, please god no! 😆 ). 

> AUX1 is also ExpressLRS lingo. Open the online documentation at expresslrs.org and search for AUX1. But really, I'm good with Channel5 or CH5 (if Channel5 doesn't fit the b&w screens).

True, but then you need to look at the **reason** it's in the lingo, and does it actually fit with the use of ELRS going beyond iNav/BF, into the wider community. i.e. can you show me many (I won't say any as I think Spektrum actually did for a short time, and then ditched that idea like a hot potato) PWM or SBUS receivers that list CH5 as AUX1? 🤪I see it as more specific lingo to those projects (and have no problem with that... just means I need to name my channels on the handset to remember which is which 🙃... which is another reason why EdgeTX is great... you can actually do that 😉).

gagarinlg commented:
How about making arming a model level function of EdgeTX
In the model config a new "Arming" config option, with following sub-configs:
- arming input (normal input selector, above 50% is considered armed)
- arming output type (message, Channel, both), where "message" triggers a module type specific message, for CRSF modules the new ELRS message, for others nothing, for now.
- arming channel, in case "Channel" or "both" is selected, defaulting to channel 5

No special function, less stuf to explain to people.

mha1 commented:
Something like this (at the moment implemented model specific). Trigger is identical to the SF Trigger and will only be visible if Arming Mode "CRSF Msg" is selected. Arming mode and Trigger can be implemented to be model specific or module specific. 

![image](https://github.com/user-attachments/assets/5f293f46-d64e-4384-b3d1-4cb75308a3e3)
![image](https://github.com/user-attachments/assets/27479a9b-1347-4a8f-8df3-b02dfee5ccde)
![image](https://github.com/user-attachments/assets/c8f8e25e-3424-46f4-a41a-bb3ad1deaf04)

gagarinlg commented:
this is close to what I had in mind. I would probably add a headline and indent the arming settings a bit, so that it is clear they belong together
This does make, at least for me, at lot more sense then a SF.

pfeerick commented:
> I was thinking (beware of the results). 

Thanks for the warning... bracing 🤪

> I propose the Arming mode setting to be along the line: `Arming: (CH5 | CRSF)` or longer` (CH5 mix | CRSF msg)`

Be warely of "mix"... as really it's an output, so CH5 should be familiar language to ETX users... Would "Telem msg" be an accurate description also? As that would remove "CRSF" from what seems to be ELRS feature?

> Only in the case of CRSF (msg) is selected a new line defining the trigger condition shall be visible. 

Or could even just make the Choice field visible on the same line since there is space. Interactivity hiding/showing adds context, minimises scrolling?

> The Trigger setting shall allow the same inputs as any SF.

Exactly.

mha1 commented:
Bracing? Duck and cover!

Finding the right terms is obviously the hardest part. CH5 is ok. Telem msg doesn't do it for me. Telem is craft to radio and that's not what it is. The radio (user) is sending the command (control). 

CRSF is referring to the communication protocol. A CRSF message is used to arm/disarm instead of a channel value. I mean it's a technicality but so is CH5. CH5 and the CRSF message are transport mediums for the same information. 

Maybe we should really call it CH5 and Switch but then also change Trigger to Switch.

I put both settings on separate lines for NV14 type radios and b&w will definitely be on two lines.

pfeerick commented:
See... I thought you would complain that telemetry was craft to handset, when in actuality it is two way, but that is splitting hairs. Changing Trigger to Switch on that screen and calling it Switch in the Choice field works for me, since that is what it is really configuring. 

I know the language/verbiage used is a hassle to pin down, but _if_ we get it right, it is mostly self explanatory without a manual. 

Whether it was one or two line made little difference, just just something to consider... Whichever fits wins 😜

mha1 commented:
I'm easy to read. CH5 and Switch it is. I'll try for one line on color radios, maybe two lines forNV14. 

How about model vs module specific? I guess you'll say module specific and I think ok.

pfeerick commented:
That would be preference, yes... Although I get what you meant with it
applying to more than one module... Meaning you would then want to check if
the other module is CRSF when changing one or the other, and configuring
that the same way also... 🤔 And here you were going "this will be easy ..
CH5 arming begone!!"?

On Fri, 1 Nov 2024, 11:07 am Michael, ***@***.***> wrote:

> I'm easy to read. CH5 and Switch it is. I'll try for one line on color
> radios, maybe two lines forNV14.
>
> How about model vs module specific? I guess you'll say module specific and
> I think ok.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2451106314>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KLH7HB3IHMS7CRIS3DZ6LH6RAVCNFSM6AAAAABQS6L54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJRGEYDMMZRGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mha1 commented:
The way it's implemented now is model specific meaning if you change the arming mode and/or switch setting in one module ie external RF and then go to the other module, ie internal RF you'd see the same setting as for the external module. And the other way around. Both module settings are synchronized because both are working on the same single source data.

Having it module specific adds one degree of freedom for the (rare) dual ELRS module case.

If your preference is the possibility to arm one by CH5, the other by CRSF message (aka Switch) I'll do it.

philmoz commented:
It should definitely be module specific data (not model). 
Having it in the model means it's taking up unnecessary space for every other RF module type.

3djc commented:
Cannot be module specific if you allow things like LS which might or not exist in said model

philmoz commented:
RF Module data inside the model, not system.
I want to avoid having this at the top level of the model data (as currently implemented).

3djc commented:
> RF Module data inside the model, not system. I want to avoid having this at the top level of the model data (as currently implemented).

Then yes, absolutely

mha1 commented:
You mean the crsf member of the ModulData union in datastructs_private.h, don't you?

![image](https://github.com/user-attachments/assets/f2bb74a5-cca5-4c88-9ef5-41e8a7d2bbb7)

mha1 commented:
This is what the UI looks after all changes. It'll fit NV 14 too, b&w will have a separate Switch line.

Is this agreeable?

![image](https://github.com/user-attachments/assets/cbe66ef3-eb58-46d8-be94-4b0080f7a927)
![image](https://github.com/user-attachments/assets/52102936-1d5e-40bd-b720-a4725b679eea)
![image](https://github.com/user-attachments/assets/3d41c0a2-b07d-4435-8bfa-f2bf964abf56)
![image](https://github.com/user-attachments/assets/88f80b46-e005-47f0-a5fb-241a0eab5bcc)
![image](https://github.com/user-attachments/assets/0da3616d-739d-44bc-a348-4c2ee09d38ad)

gagarinlg commented:
Looking good

pfeerick commented:
Perfect :)

3djc commented:
Could we have at module setup (sys/hardware), near where you setup module baudrate, an "Always armed" option ?

For Eu users who are limited to 100mW anyway, the benefits of having a un-armed state might not outweigh the risk of forgetting to arm on a model, so have a radio wide settings saying, always run ELRS armed would be nice and safe, and having now have ability to use CH5 would be great

gagarinlg commented:
Some ELRS settings are protected from changes when the ELRS stuff is armed

3djc commented:
Fair, I'll make a change for my radio only then so that it is always armed but when elrs script is running :)

froqstar commented:
> Fair, I'll make a change for my radio only then so that it is always armed but when elrs script is running :)

Please send it to me as well 😊

gagarinlg commented:
I just use the motor cut off also for ELRS arming

mha1 commented:
> I just use the motor cut off also for ELRS arming

This is the way it should be done for non-FC crafts. Link arming to something already available indicating "flight-ready". It could also be sticky logical switches on throttle > 0, gear in, tow hitch closed, or even a telemetry parameter like Alt > 10m. I personally usually setup a logical condition to start logging and use the same logical switch for arming.

If you insist on permanently on you can simply use Switch ON. The Switch selection in non-CH5 arming mode (now called Switch mode) offers the same switch option as Special Functions. Yes, you have to do it for every model but it's done in seconds.

![image](https://github.com/user-attachments/assets/a0706c43-e9ca-4484-9dbc-3db7aa9693fc)

mha1 commented:
As discussed the arming mode and switch settings are now stored module specific. Maybe there is use case for running a dual module setup and wanting to arm them independently. Anyhow, works fine. This example controls arming of the internal module by Arming mode Switch with switch SA_down (CH5 available for full-res PWM), the external module is armed by using the CH5 mixer value.

![image](https://github.com/user-attachments/assets/800935b1-2897-4df2-a043-ddf71f889bb7)


![image](https://github.com/user-attachments/assets/90cf5e3b-72f8-498c-ad77-7647c578e7ff)

mha1 commented:
New commit with changes:

- updated UI for b&w and color radios
- deleted SF ARM in favor of module confined settings
- Switch selections possible as per Special Functions
- module specific storage of Arming mode and Switch
- added missing translations (review @philmoz)
- added accidentally piece of code (review @philmoz)
- Companion partially implemented (misses Switch selection)

Tested on TX12MK2 (Sim), X9D+2019 (Sim), TX16s (HW), T15 (HW)

ToDo:
- I couldn't convince Companion to use the Special Functions type switch selection.  Need help

@elecpower Neil, can you please point me in the right direction

elecpower commented:
@mha1 Michael I'll have a look when I have a minute

elecpower commented:
@mha1 Hi Michael I have a solution ready to merge HOWEVER the model yaml file MUST store crsfArmingTrigger as a text switch not an enum as Companion has no idea how to map. Also if the model is copied to another radio the switch will be blindly mapped to the same enum which could be incorrect. Just use the existing radio yaml functions to read and write and all should be well.

Currently: SA down
crsf: 
       telemetryBaudrate: 0
       crsfArmingMode: 1
       crsfArmingTrigger: 3

Change to for SWITCH
crsf: 
       telemetryBaudrate: 0
       crsfArmingMode: 1
       crsfArmingTrigger: SA2

and for CH5
crsf: 
       telemetryBaudrate: 0
       crsfArmingMode: 0
       crsfArmingTrigger: NONE

mha1 commented:
@elecpower Hey Neil, great, thank you very much. 
I sent you an invite to collaborate. Please go ahead with comitting your solution, I'll adjust the radio side after that.

elecpower commented:
@mha1 I've pushed my commit to your repo. We can do some more testing when you have done the radio side.

mha1 commented:
> @mha1 I've pushed my commit to your repo. We can do some more testing when you have done the radio side.

Great, thank you very much. II changed the radio side and everything yaml looks good but there are some switch options missing, e.g. ON, FM. I think this is due to the wrong switch context. It looks more like the mixer context rather than the Special Function context. The radio side uses the SF context. Please check the switch selection options in the SF Tab. It has ON and FM.

elecpower commented:
> > @mha1 I've pushed my commit to your repo. We can do some more testing when you have done the radio side.
> 
> Great, thank you very much. II changed the radio side and everything yaml looks good but there are some switch options missing, e.g. ON, FM. I think this is due to the wrong switch context. It looks more like the mixer context rather than the Special Function context. The radio side uses the SF context. Please check the switch selection options in the SF Tab. It has ON and FM.

I took a punt on the filter context and surprise surprise I got it wrong. I will check the lists and fix.

mha1 commented:
@elecpower Neil, spot on. Works as it should. Many thanks.

mha1 commented:
@pfeerick @philmoz @gagarinlg @3djc @elecpower 

With Neil having finalized Companion this PR is ready for another round  of reviews. Here's the changes since the first UI implementation:

- updated b&w and color radios UIs to agreed upon UI concept without SF Arm
- updated yaml storage of Arming mode and Switch to be module specific
- deleted SF ARM in favor of module confined settings
- switch selections possible as per Special Functions
- added missing translations (review item)
- added accidentally deleted piece of code (review item)
- changed yaml storage of the arming switch to be CPN compatible (textual switch representation) 
- Companion support fully implemented (@elecpower)

I have test this on TX12MK2 (Sim), X9D+2019 (Sim), TX16s (HW), T15 (HW). HW Test were done with ELRS module and ELRS PWM receiver (and connected SBUS converter) on firmware PR https://github.comExpressLRSExpressLRS/pull/3008.

If you want to test the full system flash ELRS receiver and module using the ELRS Configurator, tab GIT PULL REQUEST, "Adds optional arming method ...".

CapnBry commented:
I endorse this addition to the regular channelspacked CRSF frame instead of creating a new type of channels packet just to include this flag. A more forward-thinking option for future expansion might be to always include the extra byte and have 0=ch5 arming, 1=switch arming, disarmed, 2=switch arming, armed. However, I do not think that is necessary (ask me about that in like 3 years when there's the question of adding something else to this frame 😅)

I really like the option being right in the module settings now. The special function was fine but now it is all on one screen and avoids the issue of the user possibly forgetting to enable the special function as well. Great that it supports logical switches too for more complex "automatic arm" logic for users who enjoy cutting off their fingers when they bump their handset!

```
TX16S internal, Tested with Betaflight
TX dynpower 25mW
333Hz
  TX set to Arming: Channel
    No CH5 mix: CH5 output 1500, never arms PASS
	CH5 mix: CH5 output is fullres, arms when CH5>1500 PASS
  TX set to Arming: Switch
	No CH5 mix: CH5 output 1500, arms when switch active PASS
    CH5 mix: CH5 output is fullres, arms when switch active PASS
      8ch is exactly 8ch (no longer 9ch) PASS
      12ch is 12ch PASS
      16ch is 16ch PASS
250Hz
  TX set to Arming: Channel
	No CH5 mix: CH5 always 1000, never arms PASS
	C5 mix: CH5 goes high if mix > 1500, arms when high PASS
    CH5 used for arming: PASS
  TX set to Arming: Switch
	No CH5 mix: CH5 output = Switch, arms when switch active PASS
	CH5 mix: CH5 output still = Switch, arms when switch active PASS

Zorro Internal, Tested with Betaflight
Tested D500, 333Hz, all above test matrix PASS

Tested with external channel dump utility
  100Hz fullres modes channel values
    8ch PASS
    12ch PASS
    16ch PASS
  50Hz
    wide switch mode PASS
    hybrid switch mode PASS

EdgeTX Extras:
Setting is stored per-model PASS
Default to CH5 arm mode PASS
```

mha1 commented:
Thanks to @CapnBry's visionary implementation to query the module's device info (CRSF message 0x29) after powering the module and his idea to use the information (module type and firmware version) it's possible to identify if an ELRS module is present and if it's ready for the added arming method.

The last commit adds presenting the arming UI extension to Internal/External RF settings only if there is an ELRS with firmware >= V4.0.0 present. The arming method options won't show if an older ELRS module with FW < V4.0.0 or any other CRSF module (TBS, GHOST) is connected.

https://github.com/user-attachments/assets/6cec6faa-b390-4b48-9703-b9e321bc1a3b

mha1 commented:
@pfeerick Translations: can I reuse `#define TR_SWITCH "Switch"` for Arming mode Switch, ie` #define TR_CRSF_ARMING_MODES "CH5",TR_SWITCH`?
Does `CH5` need to be translated?
And I'd need translations for the UI label `#define TR_ARMING_MODE "Arming"`

@ulfhedlund `#define TR_ARMING_MODE "Aktivering` is noted

ulfhedlund commented:
> @pfeerick Translations: can I reuse `#define TR_SWITCH "Switch"` for Arming mode Switch, ie` #define TR_CRSF_ARMING_MODES "CH5",TR_SWITCH`? Does `CH5` need to be translated? And I'd need translations for the UI label `#define TR_ARMING_MODE "Arming"`

In my opinion you can reuse the #define TR_SWITCH "Switch".
The "CH5" in #define TR_CRSF_ARMING_MODES "CH5", TR_SWITCH doesn't need to be translated (but it would be nice :)

pfeerick commented:
> or any other CRSF module (TBS, GHOST) is connected.

Huh? ImmersionRC is CRSF? I thought GHOST was a separate protocol type? :zany_face: 

I had hinted above https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2448429891 to use the same mechanism as binding, so it was only shown for ELRS, but was later thinking really needed to know the FW version as it wouldn't be valid for older FW, so this is much appreciated, and will useful in the future for any other version specific extensions that are added. 


> Translations: can I reuse `#define TR_SWITCH "Switch"` for Arming mode Switch, ie` #define TR_CRSF_ARMING_MODES "CH5",TR_SWITCH`?

Yup, if it is already translated, use it - don't need two copies. ;) Just need to make sure it is below TR_SWITCH. 

> Does `CH5` need to be translated?
Preferably, yes, and probably long and short form (i.e. `TR("CH5", "Channel 5"`).

mha1 commented:
@pfeerick I should've said "... doesn't show for anything else but ..."

`... but really the short form should only be be done if done elsewhere.. I have this feeling CH is hard coded and the number appended.`

Please unzip this for me.

pfeerick commented:
*Does CH5 need translation.* Only if done elsewhere. Which I  have a
suspicion is not the case. But CH5 is also the short form of Channel 5,
which *would* need translation.


On Mon, Nov 11, 2024 at 5:01 PM Michael ***@***.***> wrote:

> @pfeerick <https://github.com/pfeerick> I should've said "... doesn't
> show for anything else but ..."
>
> ... but really the short form should only be be done if done elsewhere.. I
> have this feeling CH is hard coded and the number appended.
>
> Please unzip this for me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2467383421>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KKOMZI7HX4BVDS5OYL2ABI6JAVCNFSM6AAAAABQS6L54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRXGM4DGNBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mha1 commented:
> Only if done elsewhere. Which I have a suspicion is not the case

Then let's please leave it at "CH5" which I strongly believe is understood all over world.

3djc commented:
> > Only if done elsewhere. Which I have a suspicion is not the case
> 
> Then let's please leave it at "CH5" which I strongly believe is understood all over world.

Could not agree more

![9CDEECF9-DE05-4B9E-8C15-89044507B517_1_102_o](https://github.com/user-attachments/assets/87850118-1bdf-4741-b474-3a7dc3ead9f6)

pfeerick commented:
Except we have zero control over how any given manufacturer decides to
labels things on their products, although they do generally tend to use
CH5. Consistency across *our* UI is what matters, so it should have the
same prefix as TR_CH, since it actually *is *translated.

On Mon, Nov 11, 2024 at 7:59 PM 3djc ***@***.***> wrote:

> Only if done elsewhere. Which I have a suspicion is not the case
>
> Then let's please leave it at "CH5" which I strongly believe is understood
> all over world.
>
> Could not agree more
>
> 9CDEECF9-DE05-4B9E-8C15-89044507B517_1_102_o.jpeg (view on web)
> <https://github.com/user-attachments/assets/87850118-1bdf-4741-b474-3a7dc3ead9f6>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2467715568>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KLDFAN243QEYBYYU6L2AB5WTAVCNFSM6AAAAABQS6L54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRXG4YTKNJWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mha1 commented:
Consistency is a point. TR_CH is mostly used as "CH" but there are some translations. This actually works:
`#define TR_CRSF_ARMING_MODES           TR_CH"5", TR_SWITCH`

mha1 commented:
And how about replacing the label "Arming" with "Arm using". It'd fit b&w too and might help to make the intention clear.

pfeerick commented:
That sounds good. Can be surprising how a small change makes all the
difference.

On Mon, 11 Nov 2024, 9:44 pm Michael, ***@***.***> wrote:

> And how about replacing the label "Arming" with "Arm using". It'd fit b&w
> too and might help to make the intention clear.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2467975245>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KPMP6BMXBKYEUNCSSD2ACKCRAVCNFSM6AAAAABQS6L54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRXHE3TKMRUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mha1 commented:
> That sounds good. Can be surprising how a small change makes all the difference.

Thanks, great, will do it.

mha1 commented:
@pfeerick I updated all the translations files. `CH5 `and `Switch` are now reusing TR_CH and `TR_SWITCH`, so no need to ask for translations but `#define TR_CRSF_ARMING_MODE "Arm using"` needs translating.

If you're happy with this could you please trigger the translators.

ulfhedlund commented:
SE translation:
#define TR_CRSF_ARMING_MODE TR("Aktivera m.", "Aktivera med"), unless the full length translation actually fits on BW small screen.

mha1 commented:
> SE translation: #define TR_CRSF_ARMING_MODE TR("Aktivera m.", "Aktivera med"), unless the full length translation actually fits on BW small screen.

If I count correctly b&w allows a max of 10 characters. Your b&w translation might be too long.

3djc commented:
BW doesnt use fixed width fonts, so it depends on the text itself, not char count

mha1 commented:
Translators, can we get some translations for this PR please.

`#define TR_CRSF_ARMING_MODE            "Arm using" ` is best understood as "Use arming method <choice> to arm/disarm your ELRS TX module. For more context please refer to the first post of https://github.com/ExpressLRS/ExpressLRS/pull/3008 and https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2467975245. 

- [ ] @zyren CN / TW
- [ ] @Eldenroot CZ
- [ ] @HThuren DA
- [ ] @ParkerEde / @TheIsotopes / @froqstar DE
- [ ] @Pat6874 FR
- [ ] @robustini IT
- [ ] @ToshihiroMakuuchi JP
- [x] @ulfhedlund  SE
- [ ] @ajjjjjjjj PL
- [ ] @zandorsp PT
- [ ] @kobakirill RU

froqstar commented:
DE 
`#define TR_CRSF_ARMING_MODE            "Armen via"`

@other german speakers: Do we have a german word for arming in EdgeTX/ExpressLRS yet? Personally I use "armen/gearmed" in german conversations, but not sure how widespread that is. "Entsichern" or "scharf schalten" are too long and militaristic for me.

Pat6874 commented:
FR:
"define TR_CRSF_ARMING_MODE "Mode d'armement"

Le sam. 16 nov. 2024, 20:12, Martin Mathes ***@***.***> a
écrit :

> DE
> #define TR_CRSF_ARMING_MODE "Armen via"
>
> @other <https://github.com/other> german speakers: Do we have a german
> word for arming in EdgeTX/ExpressLRS yet? Personally I use "armen/gearmed"
> in german conversations, but not sure how widespread that is. "Entsichern"
> or "scharf schalten" are too long and militaristic for me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2480742408>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6HE7XP7GN5PCKRR3EMZHZ32A6KJPAVCNFSM6AAAAABQS6L54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIOBQG42DENBQHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

robustini commented:
IT
#define TR_CRSF_ARMING_MODE            "Modo d'armamento"

3djc commented:
We also need shorter forms for BW

Pat6874 commented:
Shorter 
FR:
"Mod. Arm."


Le 17 nov. 2024, 10:55, à 10:55, 3djc ***@***.***> a écrit:
>We also need shorter forms for BW
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/EdgeTX/edgetx/pull/5641#issuecomment-2481096501
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

robustini commented:
> We also need shorter forms for BW

IT edited.

Eldenroot commented:
#define TR_CRSF_ARMING_MODE            "Arm mód"

zyren commented:
******************************
CN
```c
#define TR_CRSF_ARMING_MODE            "解锁类型" 
```
********************************
TW
```c
#define TR_CRSF_ARMING_MODE            "解鎖類型" 
```
********************************

mha1 commented:
Thank you guys. So far I have:

CN: #define TR_CRSF_ARMING_MODE "解锁类型" 
TW: #define TR_CRSF_ARMING_MODE "解鎖類型"
CZ: #define TR_CRSF_ARMING_MODE "Arm mód"
IT: #define TR_CRSF_ARMING_MODE TR("Modo arm", "Modo d'armamento")
FR: #define TR_CRSF_ARMING_MODE TR("Mod. Arm.", "Mode d'armement")
DE: #define TR_CRSF_ARMING_MODE "Armen via"

kobakirill commented:
#define TR_CRSF_ARMING_MODE            "Arm режим"

HThuren commented:
DA translation:
#define TR_CRSF_ARMING_MODE TR("Aktiver m.", "Aktiver med")

Luznatural commented:
ES translation:
#define TR_CRSF_ARMING_MODE TR "Modo Armado"

zandorsp commented:
PT
#define TR_CRSF_ARMING_MODE TR("Arm usd.", "Armar usando")

ToshihiroMakuuchi commented:
JP
#define TR_CRSF_ARMING_MODE            "アーム ロック解除"

mha1 commented:
So far only FI, HE, NL and PL are missing. Does anyone know who to contact for FI, HE, NL?

pfeerick commented:
HE translator asked not to be pinged for each PR, I think they plan on doing  translations nearer to a release as a batch. For the others, nada, nope... so until someone says "I speak the lingo, I'll do it"... they get ignored :innocent:

mha1 commented:
> HE translator asked not to be pinged for each PR, I think they plan on doing translations nearer to a release as a batch. For the others, nada, nope... so until someone says "I speak the lingo, I'll do it"... they get ignored 😇

Alright, then we're through
